### ISSUE TITLE
Handle multiple authentication methods is not correct.

### ISSUE THREAD

**Issue by davecramer** [r2dbc/r2dbc-postgresql#19]
The way the protocol works is that the backend sends a message requesting the authentication it desires
After the auth request, the next 4 bytes are the authentication method. I think what needs to be done is to have multiple handlers added to the startup message and then react to the correct authentication method ?

**Comment by nebhale** [comment]
Looking at [51.1](https://www.postgresql.org/docs/9.6/static/protocol-flow.html), we [send the Startup Message](https://github.com/r2dbc/r2dbc-postgresql/blob/master/src/main/java/io/r2dbc/postgresql/client/StartupMessageFlow.java#L60) and then [handle the resulting Authentication Message](https://github.com/r2dbc/r2dbc-postgresql/blob/master/src/main/java/io/r2dbc/postgresql/client/StartupMessageFlow.java#L66) should one be returned.  For the PoC, I only needed a small handful of possible authentication methods, and we do some [limited selection](https://github.com/r2dbc/r2dbc-postgresql/blob/master/src/main/java/io/r2dbc/postgresql/authentication/PasswordAuthenticationHandler.java#L49) for the correct ones.  This [handler mechanism](https://github.com/r2dbc/r2dbc-postgresql/blob/master/src/main/java/io/r2dbc/postgresql/PostgresqlConnectionFactory.java#L82) needs to be expanded to support other authentication mechanism (like #6).

**Comment by davecramer** [comment]
[51.1](https://www.postgresql.org/docs/9.6/static/protocol-flow.html) does not tell us in advance what authentication method will be I think we need to handle it directly in the [startupmessage](https://github.com/r2dbc/r2dbc-postgresql/blob/f207ea7602505d9459569defcf207e90bcc73ed7/src/main/java/io/r2dbc/postgresql/client/StartupMessageFlow.java#L66)
This would require some refactoring to not preset the authentication handler in https://github.com/r2dbc/r2dbc-postgresql/blob/f207ea7602505d9459569defcf207e90bcc73ed7/src/main/java/io/r2dbc/postgresql/client/StartupMessageFlow.java#L51 , but we will need the username and password (Possibly more ?)

**Comment by nebhale** [comment]
Correct.  Effectively, the `StartupMessageFlow` needs a provider to dynamically determine at runtime.  I've [started the refactoring](https://github.com/r2dbc/r2dbc-postgresql/commit/e5b9141a40d9720d1a88f3170954c32135572d00), and now all we need to do is create `AuthenticationHandlers` for the rest of the authentication methods we would like to support.

**Comment by davecramer** [comment]
I had to add

diff --git a/src/main/java/io/r2dbc/postgresql/PostgresqlConnection.java
b/src/main/java/io/r2dbc/postgresql/PostgresqlConnection.java
index 40d31af..6dfd1ac 100644
--- a/src/main/java/io/r2dbc/postgresql/PostgresqlConnection.java
+++ b/src/main/java/io/r2dbc/postgresql/PostgresqlConnection.java
@@ -23,6 +23,7 @@ import io.r2dbc.postgresql.client.TransactionStatus;
 import io.r2dbc.postgresql.codec.Codecs;
 import io.r2dbc.spi.Connection;
 import io.r2dbc.spi.IsolationLevel;
+import io.r2dbc.spi.Mutability;
 import org.reactivestreams.Publisher;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -214,4 +215,8 @@ public final class PostgresqlConnection implements
Connection {
         return Mono.defer(() ->
Mono.just(f.apply(this.client.getTransactionStatus())));
     }

+    @Override
+    public Publisher<Void> setTransactionMutability(Mutability mutability)
{
+        return Mono.empty();
+    }
 }
to get your refactoring to compile...
Dave Cramer
